### ISSUE TITLE
Upgrader: php 8 issues upgrading old 1.0 forums

### ISSUE THREAD

**Issue by sbulen** [SimpleMachines/SMF#7081]
#### Description
The upgrader runs fine in php7.4, but reports lots of errors in php8.  A sampling:
```
	Line 80:  +++  Recounting personal message totals...Error: Undefined array key "m" File: yada\yada\upgrade.php(2155) : eval()'d code Line: 10
	Line 80:  +++  Recounting personal message totals...Error: Undefined array key "m" File: yada\yada\upgrade.php(2155) : eval()'d code Line: 10
	Line 86:  +++  Converting topic notifications...Error: Undefined array key "t" File: yada\yada\upgrade.php(2155) : eval()'d code Line: 9 done.
```

Many undefined index errors converting old themes as well.  

Note 1.1 runs fine.  This is specific to 1.0.  

The upgrade runs to completion.  I'd consider this a pretty low priority.  Would likely impact 2.0 as well - they use almost identical scripts.

### Steps to reproduce
1. Upgrade a smf 1.0 forum to 2.1 using php8

### Environment (complete as necessary)
- Version/Git revision: Current
- Database Type: mysql
- Database Version: 5.7
- PHP Version: php 8

**Comment by albertlast** [comment]
Is this a supported upgrade path?
The 1.0 forum doesn't run with php 8.0 i guess.

**Comment by sbulen** [comment]
Yes.

**Comment by albertlast** [comment]
1.0 is not supported any more,
so a simple yes sounds wrong.

**Comment by Sesquipedalian** [comment]
*Running* SMF 1.0 is unsupported. *Upgrading from* SMF 1.0 is supported. After all, if someone asks for support for SMF 1.0, how can we legitimately tell them that they should upgrade to SMF 2.1 if it is impossible to do so?

**Comment by Oldiesmann** [comment]
Theoretically you should even be able to upgrade all the way from YaBB SE at this point, though you'd likely run into a lot of other problems along the way given how much PHP and MySQL have changed in 17 years...

**Comment by sbulen** [comment]
I test YaBB SE along with SMF 1.0, 1.,1 and 2.0 upgrades when I run a round of upgrade tests.  That's how I found this issue.  This was the first time I did a set using php8; previous tests were 7.4.  Depending where I run the tests, they're using mysql 5.6 or 5.7.  

When issues are found, we've been fixing the 2.1 & the 2.0 upgrade scripts, so they are ~in sync.   -ish.

The YaBB upgrade only works for that last officially released version (1.5.5).  Nothing older, nothing newer...  But they really do work fine - certainly on fairly vanilla forums.  (Check out the last post date:  https://shawnbulen.com/yabbse/index.php)

It's getting harder to keep it all afloat though.  2.1 should probably be the last upgrader to reach back that far.

**Comment by sbulen** [comment]
The problem here is that the error suppression operator, @, behaves very differently in php8+.  In the past, all errors were suppressed.  Now, it will no longer be silent on 'fatal' errors.  

And it appears 'undefined array key' is considered fatal...  

So, a lot of code that looks like this now errors out, where it didn't previously:
`  $a = @$GLOBALS['george'];`

I believe we can address this by making the upgrader error handler a little smarter.  

More here:
https://php.watch/versions/8.0/fatal-error-suppression

The types of things that are erroring out now include:
https://github.com/SimpleMachines/SMF2.1/blob/7d5b3a6356ecf45110a456508bafb47e2bf78655/other/upgrade.php#L1882
https://github.com/SimpleMachines/SMF2.1/blob/7d5b3a6356ecf45110a456508bafb47e2bf78655/other/upgrade_1-0.sql#L1129
